### PR TITLE
Allow muting /sound per client

### PR DESCRIPTION
**Pull request by bratkartoff**
Additions:
- "mute" / "unmute" entry in the lobby (the context menu for the client)
- `/mute [client]` and `/unmute [client]` chat commands
- speaker button in the Network (ingame) window
- new config value "MuteSoundChatCommand" that determines whether clients are muted by default

### CODE REVIEW

File: src/C4Console.cpp [hunk 1]
```diff
@@ -598,7 +598,7 @@ bool C4Console::In(const char *szText)
 		// done
 		return true;
 	}
-	// begins with '#'? then it's a message. Route cia ProcessInput to allow #/sound
+	// begins with '#'? then it's a message. Route via ProcessInput to allow #/sound
```
Comment (maxmitti) [inline src/C4Console.cpp:601]: This has nothing to do with this PR. It could be in an extra spelling-fixes-commit.

File: src/C4PlayerInfoListBox.cpp [hunk 2]
```diff
@@ -921,21 +921,25 @@ C4GUI::ContextMenu *C4PlayerInfoListBox::ClientListItem::OnContext(C4GUI::Elemen
 	C4Client *pClient = GetClient();
 	// create context menu
 	C4GUI::ContextMenu *pMenu = new C4GUI::ContextMenu();
+	// helper function
+	auto AddMenuItem = [&](const char *text, const char *description, auto callbackFn)
```
Comment (maxmitti) [inline src/C4PlayerInfoListBox.cpp:925]: I am not sure about the (still unpublished) style guidelines, but I prefer explicitly specifying captured variables.

File: src/C4PlayerInfoListBox.cpp [hunk 3]
```diff
@@ -921,21 +921,25 @@ C4GUI::ContextMenu *C4PlayerInfoListBox::ClientListItem::OnContext(C4GUI::Elemen
...
+	{
+		pMenu -> AddItem(LoadResStr(text), LoadResStr(description), C4GUI::Ico_None,
```
Comment (maxmitti) [inline src/C4PlayerInfoListBox.cpp:927]: No spaces around `->`

File: src/C4PlayerInfoListBox.cpp [hunk 4]
```diff
@@ -921,21 +921,25 @@ C4GUI::ContextMenu *C4PlayerInfoListBox::ClientListItem::OnContext(C4GUI::Elemen
...
 	// info
-	StdStrBuf strClientInfoDesc(LoadResStr("IDS_NET_CLIENTINFO_DESC"));
-	pMenu->AddItem(LoadResStr("IDS_NET_CLIENTINFO"), strClientInfoDesc.getData(), C4GUI::Ico_None,
-		new C4GUI::CBMenuHandler<ClientListItem>(this, &ClientListItem::OnCtxInfo));
+	AddMenuItem("IDS_NET_CLIENTINFO", "IDS_NET_CLIENTINFO_DESC", &ClientListItem::OnCtxInfo);
+	// mute / unmute
+	if (pClient->isMuted())
+		AddMenuItem("IDS_NET_UNMUTE", "IDS_NET_UNMUTE_DESC", &ClientListItem::OnCtxUnmute);
+	else
+		AddMenuItem("IDS_NET_MUTE", "IDS_NET_MUTE_DESC", &ClientListItem::OnCtxMute);
```
Comment (maxmitti) [inline src/C4PlayerInfoListBox.cpp:942]: The (still unpublished) style guidelines allow omitting braces only for ifs without else.
However this might be irrelevant due to other changes.

**Review by maxmitti — CHANGES_REQUESTED** [review]
Overall, this looks nice. However there are some minor changes needed.

I think the ClientListItem::OnCtx{Unmute,Mute} callbacks could be merged into a single one (like for (de)activating).
As a result, the context menu entry and the button could share a toggling function.

File: .ccls [hunk 5]
```diff
@@ -0,0 +1,3 @@
+clang
+%c
+-Isrc
```
Comment (Fulgen301) [inline .ccls]: What is this? It doesn't seem to be related to this commit or the PR.

File: src/C4MessageInput.cpp [hunk 6]
```diff
@@ -419,6 +419,8 @@ bool C4MessageInput::ProcessCommand(const char *szCommand)
 		LogF("/team [message] - %s", LoadResStr("IDS_MSG_SENDAPRIVATEMESSAGETOYOUR"));
 		LogF("/me [action] - %s", LoadResStr("IDS_TEXT_PERFORMANACTIONINYOURNAME"));
 		LogF("/sound [sound] - %s", LoadResStr("IDS_TEXT_PLAYASOUNDFROMTHEGLOBALSO"));
+		LogF("/mute [client] - %s", LoadResStr("IDS_NET_MUTE_DESC"));
+		LogF("/unmute [client] - %s", LoadResStr("IDS_NET_UNMUTE_DESC"));
```
Comment (Fulgen301) [inline src/C4MessageInput.cpp:423]: The string table entries should have the same format as the other help entries.

File: src/C4MessageInput.cpp [hunk 7]
```diff
@@ -649,6 +651,22 @@ bool C4MessageInput::ProcessCommand(const char *szCommand)
 		return true;
 	}
 
+	// mute
+	if (SEqual(szCmdName, "mute"))
+	{
+		if (auto *client = Game.Clients.getClientByName(pCmdPar))
```
Comment (Fulgen301) [inline src/C4MessageInput.cpp:657]: This creates a warning - I recommend changing it to `if (auto *client = Game.Clients.getClientByName(pCmdPar); client)`

File: src/C4MessageInput.cpp [hunk 8]
```diff
@@ -649,6 +651,22 @@ bool C4MessageInput::ProcessCommand(const char *szCommand)
...
+			client->SetMuted(true);
+		return true;
+	}
+
+	// unmute
+	if (SEqual(szCmdName, "unmute"))
+	{
+		if (auto *client = Game.Clients.getClientByName(pCmdPar))
```
Comment (Fulgen301) [inline src/C4MessageInput.cpp:665]: Same as above.

**Review by Fulgen301 — CHANGES_REQUESTED** [review]

File: src/C4Config.h [hunk 9]
```diff
@@ -156,6 +156,7 @@ class C4ConfigSound
 	int32_t MusicVolume;
 	int32_t SoundVolume;
 	std::chrono::seconds SoundCommandCooldown;
+	bool MuteSoundChatCommand; // whether to mute /sound by default
```
Comment (Fulgen301) [inline src/C4Config.h:159]: The existing configuration value uses `SoundCommand` instead of `SoundChatCommand` - this should be consistent.

**Review by Fulgen301 — CHANGES_REQUESTED** [review]

File: src/C4Client.cpp [hunk 10]
```diff
@@ -85,10 +85,10 @@ void C4ClientCore::CompileFunc(StdCompiler *pComp)
 // *** C4Client
 
 C4Client::C4Client()
-	: fLocal(false), pNetClient(nullptr), last_lobby_ready_change(0) {}
+	: C4Client(C4ClientCore()) {}
 
 C4Client::C4Client(const C4ClientCore &Core)
-	: Core(Core), fLocal(false), pNetClient(nullptr), pNext(nullptr), last_lobby_ready_change(0) {}
+	: Core(Core), fLocal(false), pNetClient(nullptr), pNext(nullptr), last_lobby_ready_change(0), muted(Config.Lobby.MuteSoundCommand) {}
```
Comment (Fulgen301) [inline src/C4Client.cpp]: Style guide mandates brace initialization.

File: src/C4Control.cpp [hunk 11]
```diff
@@ -1154,12 +1154,16 @@ void C4ControlMessage::Execute() const
 	break;
 
 	case C4CMT_Sound:
+	{
 		// tehehe, sound!
-		if (StartSoundEffect(szMessage, false, 100, nullptr))
+		C4Client* pClient = Game.Clients.getClientByID(iByClient);
+		if (pClient->isMuted() || StartSoundEffect(szMessage, false, 100, nullptr))
```
Comment (Fulgen301) [inline src/C4Control.cpp]: This can be shortened by using if-init.

File: src/C4PlayerInfoListBox.cpp (hunk 2 above)
Comment (Fulgen301) [reply src/C4PlayerInfoListBox.cpp:925]: Agreed.

**Review by Fulgen301 — CHANGES_REQUESTED** [review]

File: src/C4GameLobby.cpp [hunk 12]
```diff
@@ -602,6 +602,8 @@ C4GUI::Edit::InputResult MainDlg::OnChatInput(C4GUI::Edit *edt, bool fPasting, b
 				LogF("/observer [client] - %s", LoadResStr("IDS_TEXT_SETTHESPECIFIEDCLIENTTOOB"));
 				LogF("/me [action] - %s", LoadResStr("IDS_TEXT_PERFORMANACTIONINYOURNAME"));
 				LogF("/sound [sound] - %s", LoadResStr("IDS_TEXT_PLAYASOUNDFROMTHEGLOBALSO"));
+				LogF("/mute [client] - %s", LoadResStr("IDS_NET_MUTE_DESC"));
+				LogF("/unmute [client] - %s", LoadResStr("IDS_NET_UNMUTE_DESC"));
```
Comment (Fulgen301) [inline src/C4GameLobby.cpp:606]: The string table entries should have the same format as the other help entries.

File: src/C4Client.cpp [hunk 13]
```diff
@@ -85,10 +85,10 @@ void C4ClientCore::CompileFunc(StdCompiler *pComp)
 // *** C4Client
 
 C4Client::C4Client()
-	: fLocal(false), pNetClient(nullptr), last_lobby_ready_change(0), lastSound{} {}
+	: C4Client(C4ClientCore()) {}
```
Comment (Fulgen301) [inline src/C4Client.cpp:88]: Why did you change that? Also, style guide mandates brace initialization.

File: src/C4PlayerInfoListBox.cpp [hunk 14]
```diff
@@ -963,6 +967,20 @@ void C4PlayerInfoListBox::ClientListItem::OnCtxInfo(C4GUI::Element *pListItem)
...
+	if (pClient)
+		pClient->SetMuted(false);
+}
+
+void C4PlayerInfoListBox::ClientListItem::OnCtxMute(C4GUI::Element *pListItem)
+{
+	C4Client *pClient = GetClient();
+	if (pClient)
+		pClient->SetMuted(true);
+}
```
Comment (Fulgen301) [inline src/C4PlayerInfoListBox.cpp:982]: This can use if-init as well.

File: src/C4PlayerInfoListBox.cpp [hunk 15]
```diff
@@ -921,21 +921,25 @@ C4GUI::ContextMenu *C4PlayerInfoListBox::ClientListItem::OnContext(C4GUI::Elemen
 	C4Client *pClient = GetClient();
 	// create context menu
 	C4GUI::ContextMenu *pMenu = new C4GUI::ContextMenu();
+	// helper function
+	auto AddMenuItem = [&](const char *text, const char *description, auto callbackFn)
+	{
+		pMenu -> AddItem(LoadResStr(text), LoadResStr(description), C4GUI::Ico_None,
+			new C4GUI::CBMenuHandler<ClientListItem>(this, callbackFn));
```
Comment (Fulgen301) [inline src/C4PlayerInfoListBox.cpp:928]: Style guide mandates brace initialization.

File: src/C4Network2Dialogs.h [hunk 16]
```diff
@@ -119,7 +119,9 @@ class C4Network2ClientListBox : public C4GUI::ListBox
 		C4GUI::Icon *pStatusIcon;  // client network status
 		C4GUI::Label *pName;       // client name
 		C4GUI::Label *pPing;       // client control ping
-		C4GUI::IconButton *pActivateBtn, *pKickBtn; // buttons for host
+		C4GUI::IconButton *pMuteBtn;     // mute client (/sound command)
```
Comment (Fulgen301) [inline src/C4Network2Dialogs.h:122]: Hungarian notation is forbidden by the style guide, however, I'm not sure whether it's better to keep in line here or not.

File: src/C4Network2Dialogs.cpp [hunk 17]
```diff
@@ -238,6 +242,15 @@ const C4Client *C4Network2ClientListBox::ClientListItem::GetClient() const
 	return Game.Clients.getClientByID(iClientID);
 }
 
+void C4Network2ClientListBox::ClientListItem::OnButtonMute(C4GUI::Control *pButton)
+{
+    C4Client* pClient = Game.Clients.getClientByID(iClientID);
+	if (pClient)
+	{
+		pClient->SetMuted(!pClient->isMuted());
+	}
```
Comment (Fulgen301) [inline src/C4Network2Dialogs.cpp:251]: Indentation errors. Also, you can use if-init.

File: src/C4Network2Dialogs.cpp [hunk 18]
```diff
@@ -136,16 +136,18 @@ C4Network2ClientListBox::ClientListItem::ClientListItem(class C4Network2ClientLi
 	}
 	pName = new C4GUI::Label(sNameLabel.getData(), iIconSize + IconLabelSpacing, iVerticalIndent, ALeft);
 	int iPingRightPos = GetBounds().Wdt - IconLabelSpacing;
-	if (Game.Network.isHost()) iPingRightPos -= 48;
+	pMuteBtn = new C4GUI::CallbackButtonEx<C4Network2ClientListBox::ClientListItem, C4GUI::IconButton>(C4GUI::Ico_Sound, GetToprightCornerRect((std::max)(iIconSize, 16), (std::max)(iIconSize, 16), 2, 1, 0), 0, this, &ClientListItem::OnButtonMute);
```
Comment (Fulgen301) [inline src/C4Network2Dialogs.cpp:139]: Style guide mandates brace initialization.

File: src/C4Network2Dialogs.cpp [hunk 19]
```diff
@@ -108,7 +108,7 @@ void C4Network2ClientDlg::UpdateText()
 // C4Network2ClientListBox::ClientListItem
 
 C4Network2ClientListBox::ClientListItem::ClientListItem(class C4Network2ClientListBox *pForDlg, int iClientID)
-	: ListItem(pForDlg, iClientID), pStatusIcon(nullptr), pName(nullptr), pPing(nullptr), pActivateBtn(nullptr), pKickBtn(nullptr)
+	: ListItem(pForDlg, iClientID), pStatusIcon(nullptr), pName(nullptr), pPing(nullptr), pMuteBtn(nullptr), pActivateBtn(nullptr), pKickBtn(nullptr)
```
Comment (Fulgen301) [inline src/C4Network2Dialogs.cpp:111]: Style guide mandates brace initialization.

File: src/C4Client.h [hunk 20]
```diff
@@ -117,6 +119,8 @@ class C4Client
 	void SetLocal();
 	void ResetSoundCooldown() { lastSound = std::chrono::steady_clock::now(); }
 
+	void SetMuted(bool muted) { this->muted = muted; }
```
Comment (Fulgen301) [inline src/C4Client.h:122]: I would actually remove that empty line.

File: src/C4Client.cpp [hunk 21]
```diff
@@ -85,10 +85,10 @@ void C4ClientCore::CompileFunc(StdCompiler *pComp)
...
 
 C4Client::C4Client(const C4ClientCore &Core)
-	: Core(Core), fLocal(false), pNetClient(nullptr), pNext(nullptr), last_lobby_ready_change(0), lastSound{} {}
+	: Core(Core), fLocal(false), pNetClient(nullptr), pNext(nullptr), last_lobby_ready_change(0), muted(Config.Sound.MuteSoundChatCommand), lastSound{} {}
```
Comment (Fulgen301) [inline src/C4Client.cpp:91]: Style guide mandates brace initialization.

**Review by Fulgen301 — CHANGES_REQUESTED** [review]

File: src/C4Client.cpp (hunk 13 above)
Comment (maxmitti) [reply src/C4Client.cpp:88]: It actually makes sense, instead of initializing muted here too.

File: src/C4PlayerInfoListBox.h [hunk 22]
```diff
@@ -205,8 +205,7 @@ class C4PlayerInfoListBox : public C4GUI::ListBox
 		void OnCtxKick(C4GUI::Element *pListItem); // kick item selected in client ctx menu
 		void OnCtxActivate(C4GUI::Element *pListItem); // toggle player/observer
 		void OnCtxInfo(C4GUI::Element *pListItem); // show info dlg (modal)
-		void OnCtxUnmute(C4GUI::Element *pListItem); // unmute /sound
-		void OnCtxMute(C4GUI::Element *pListItem); // mute /sound
+		void OnCtxToggleMute(C4GUI::Element *pListItem); // mute /sound
```
Comment (Fulgen301) [inline src/C4PlayerInfoListBox.h:208]: That comment is now incorrect, as it toggles the mute status.

**Review by Fulgen301 — APPROVED** [review]
